client/consensus/manual-seal/src/heartbeat_stream.rs

```diff
@@ -0,0 +1,149 @@
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Heartbeat stream is a stream adapter built around the manual-seal `EngineCommand` stream.
+//! It adds additional logic of creating heartbeat block and having a cooldown period.
+
+use std::{pin::Pin, time::{Duration, Instant}};
+use futures::{
+	prelude::*,
+	task::{Context, Poll},
+};
+use futures_timer::Delay;
+use crate::rpc::EngineCommand;
+
+/// Options to manage the behavior of the heartbeat stream
+struct Options {
+	// The amount of time passed that a new empty block will be generated when no transactions
+	// in the tx pool.
+	heartbeat: Option<Duration>,
+	// The cooldown duration after a block has been authored.
+	cooldown: Option<Duration>,
+	// Whether the created block is finalized of not.
+	finalize: bool,
+}
+
+/// Heartbeat stream is a stream adapter built around the manual-seal `EngineCommand` stream.
+/// It adds additional logic of creating heartbeat block and having a cooldown period.
+pub struct HeartbeatStream<Hash> {
+	// The `EngineCommand` stream passed from the caller
+	command_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	// Delay future to control when to wake up next
+	delay_for: Option<Delay>,
+	// To remember when the last block is generated
+	last_blocktime: Option<Instant>,
+	// Heartbeat options
+	options: Options,
+}
+
+impl<Hash> HeartbeatStream<Hash> {
+	pub fn new(
+		command_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+		heartbeat: Option<Duration>,
+		cooldown: Option<Duration>,
+		finalize: bool
+	) -> Result<Self, &'static str> {
+		match (heartbeat, cooldown) {
+			(Some(heartbeat), Some(cooldown)) if cooldown > heartbeat =>
+				Err("`cooldown` must not be larger than the `heartbeat`, if they are both set."),
+			_ => Ok(()),
+		}?;
+
+		let delay_for = heartbeat.and_then(|hb| Some(Delay::new(hb)));
+		Ok(Self {
+			command_stream,
+			delay_for,
+			last_blocktime: None,
+			options: Options { heartbeat, cooldown, finalize }
+		})
+	}
+
+	fn create_block_now_and_reset_delay(&mut self) {
+		self.last_blocktime = Some(Instant::now());
+		self.delay_for = self.options.heartbeat.and_then(|hb| Some(Delay::new(hb)));
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```

Comment:
The logic in this function is too complicated in here.

There are also some questions that arise, what happens when we drop such an engine command? Like we should finalize a block and now it is just dropped?

And why didn't you followed my earlier proposal for this? That would not result in dropping any events.